tests/e2e/e2e_test.go

```diff
@@ -0,0 +1,268 @@
+	s.Require().Len(whitelistedAddresses, 2)
+	s.Require().Contains(whitelistedAddresses, whitelistAddr1)
+	s.Require().Contains(whitelistedAddresses, whitelistAddr2)
+}
+
+func (s *IntegrationTestSuite) TestFeeTax() {
+	chain := s.configurer.GetChainConfig(0)
+	node, err := chain.GetDefaultNode()
+	s.Require().NoError(err)
+
+	transferAmount1 := sdkmath.NewInt(20000000)
+	transferCoin1 := sdk.NewCoin("uluna", transferAmount1)
+
+	validatorAddr := node.GetWallet(initialization.ValidatorWalletName)
+	s.Require().NotEqual(validatorAddr, "")
+
+	validatorBalance, err := node.QuerySpecificBalance(validatorAddr, "uluna")
+	s.Require().NoError(err)
+
+	test1Addr := node.CreateWallet("test1")
+
+	// Test 1: banktypes.MsgSend
+	// burn tax with bank send
+	node.BankSend(transferCoin1.String(), validatorAddr, test1Addr)
+
+	subAmount := sdk.NewDecFromInt(transferAmount1).Mul(sdk.NewDecWithPrec(102, 2)).TruncateInt()
+	taxRate, err := node.QueryTaxRate()
+	s.Require().NoError(err)
+	s.Require().Equal(subAmount, taxRate.Add(sdk.OneDec()).MulInt(transferAmount1).TruncateInt())
+
+	decremented := validatorBalance.Sub(sdk.NewCoin("uluna", subAmount))
+	newValidatorBalance, err := node.QuerySpecificBalance(validatorAddr, "uluna")
+	s.Require().NoError(err)
+
+	balanceTest1, err := node.QuerySpecificBalance(test1Addr, "uluna")
+	s.Require().NoError(err)
+
+	s.Require().Equal(balanceTest1.Amount, transferAmount1)
+	s.Require().Equal(newValidatorBalance, decremented)
+
+	// Test 2: try bank send with grant
+	test2Addr := node.CreateWallet("test2")
+	transferAmount2 := sdkmath.NewInt(10000000)
+	transferCoin2 := sdk.NewCoin("uluna", transferAmount2)
+
+	node.BankSend(transferCoin2.String(), validatorAddr, test2Addr)
+	node.GrantAddress(test2Addr, test1Addr, transferCoin2.String(), "test2")
+
+	validatorBalance, err = node.QuerySpecificBalance(validatorAddr, "uluna")
+	s.Require().NoError(err)
+
+	node.BankSendFeeGrantWithWallet(transferCoin2.String(), test1Addr, validatorAddr, test2Addr, "test1")
+
+	newValidatorBalance, err = node.QuerySpecificBalance(validatorAddr, "uluna")
+	s.Require().NoError(err)
+
+	balanceTest1, err = node.QuerySpecificBalance(test1Addr, "uluna")
+	s.Require().NoError(err)
+
+	balanceTest2, err := node.QuerySpecificBalance(test2Addr, "uluna")
+	s.Require().NoError(err)
+
+	s.Require().Equal(balanceTest1.Amount, transferAmount1.Sub(transferAmount2))
+	s.Require().Equal(newValidatorBalance, validatorBalance.Add(transferCoin2))
+	s.Require().Equal(balanceTest2.Amount, sdk.NewDecWithPrec(98, 2).MulInt(transferAmount2).TruncateInt())
+
+	// Test 3: banktypes.MsgMultiSend
+	validatorBalance, err = node.QuerySpecificBalance(validatorAddr, "uluna")
+	s.Require().NoError(err)
+
+	node.BankMultiSend(transferCoin1.String(), false, validatorAddr, test1Addr, test2Addr)
+
+	newValidatorBalance, err = node.QuerySpecificBalance(validatorAddr, "uluna")
+	s.Require().NoError(err)
+
+	subAmount = sdk.NewDecWithPrec(204, 2).MulInt(transferAmount1).TruncateInt()
```

Comment:
I noticed that you guys have these weird exact values of 204, 98, 102 all over the test. 

Provided that a process is a black box f(x1, x2, x3) which is correct and deterministic. Given fixed value (x1, x2, x3) as inputs, it should return the same output (y1, y2, y3)

In this case, since you guys have these weird value from start to end of this test function, it is like f_a1,a2,a3(x1, x2, x3). My reasoning is that if I change (a1, a2, a3), then with fixed (x1, x2, x3), it will return different (y1, y2, y3). It could also be that my reasoning is not fully.

Preferably it should look like this one with changable values defined at the start as inputs: https://github.com/classic-terra/core/blob/main/custom/auth/ante/fee_test.go#L720

Or, maybe with comments of how these calculations are formulated

Reply:
@StrathCole you're right. We will convert it into a constant and perform the mathematical operations beneath it.